README.md
pornhub
=======

Access [pornhub](http://www.pornhub.com/) (WARNING: NSFW) programmatically with
Node.JS. Uses [hubtraffic.com](http://http://www.hubtraffic.com/) (go signup!)

Overview
--------

This module scrapes the HTML of pornhub.com and gives you information you can
use in your node programs!

Super Quickstart
----------------

```javascript
var ph = require("pornhub");

ph.details("http://www.pornhub.com/view_video.php?viewkey=591533139", function(err, details) {
  console.log(err, details);
});
```

Installation
------------

Available via [npm](http://npmjs.org/):

> $ npm install pornhub

Or via git:

> $ git clone git://github.com/electblake/pornhub.git node_modules/pornhub

Usage
-----

The `pornhub` package exports a single documented function right now, `details`.

pornhub API
-----------

**details**

Gets details about a specific video.

```javascript
ph.details(uri, cb);
```

```javascript
// get details about a video
ph.details("http://www.pornhub.com/view_video.php?viewkey=591533139", function(err, details) {
  console.log(err, details);
});
```

Arguments

* _uri_ - a full url to an pornhub video page
* _cb_ - a callback to be called in the normal node way with an error (or null)
  and the video details

**search**

Performs a search and returns the results. `parameters` is an object that is
used to construct the query string for the search URL.

Available parameters are:

- `category` (Optional)

- `page` (Optional) Integer

- `search` (Optional) Text

- `stars` (Optional) Array

- `tags` (Optional) Array

- `thumbsize` (Required). Possible values are small,medium,large,small_hd,medium_hd,large_hd

signup for [hubtraffic.com](http://http://www.hubtraffic.com/) and see method `searchVideos` for details

```javascript
ph.search(parameters, cb);
```

```javascript
// get details about a video
ph.search({search: "hard", tags: ["teen"], thumbsize: "medium"}, function(err, results) {
  console.log(err, results);
});
```

Arguments

* _parameters_ - an object containing parameters for the query string (see above
  for details about its properties)
* _cb_ - a callback to be called in the normal node way with an error (or null)
  and the search results

Example
-------

Also see [example.js](https://github.com/electblake/pornhub/blob/master/example.js).

```javascript
#!/usr/bin/env node

var xv = require("pornhub");

// get details about a video
xv.details("http://www.pornhub.com/video3823160/stranded_busty_sweetie_decides_to_have_sex_with_a_stranger_in_public_for_money", function(err, details) {
  console.log(err, details);
});
```

Output (example):

```
null { title: 'Stranded busty sweetie decides to have sex with a stranger in public for money',
  duration: '5 min',
  tags:
   [ 'amateur',
     'bigtits',
     'brunette',
     'busty',
     'european',
     'flashing',
     'hardcore',
     'money',
     'outdoor',
     'POV',
     'public' ],
  flv: 'http://porn.im.cbce7ca3.3823160.x.pornhub.com/videos/flv/f/e/6/pornhub.com_fe6e41a62c4162b6b9ff750620acf599.flv?e=1366800046&ri=1024&rs=85&h=2cf1096659393b4c153b7cdfe93a8889',
  thumb: 'http://img-l3.pornhub.com/videos/thumbslll/fe/6e/41/fe6e41a62c4162b6b9ff750620acf599/fe6e41a62c4162b6b9ff750620acf599.23.jpg',
  html: '<iframe src="http://flashservice.pornhub.com/embedframe/3823160" frameborder=0 width=510 height=400 scrolling=no></iframe>' }
```

License
-------

3-clause BSD. A copy is included with the source.

Contact
-------

* GitHub ([electblake](http://github.com/electblake))
* Twitter ([@electblake](http://twitter.com/electblake))


CHANGELOG.md
### 0.0.5 (2015-09-03)


### 0.0.4 (2015-09-03)


### 0.0.3 (2015-09-03)


#### Features

* **details:** added details and search in xvideos format ([d1c4c1e2](https://github.com/electblake/pornhub/commit/d1c4c1e218ca1a238914d6a5cd4b3bb2af9a5d8b))


### 0.0.2 (2015-09-03)


#### Features

* **details:** added details and search in xvideos format ([d1c4c1e2](https://github.com/electblake/pornhub/commit/d1c4c1e218ca1a238914d6a5cd4b3bb2af9a5d8b))



pornhub.js
var cheerio = require("cheerio"),
    http = require("http"),
    qs = require("querystring"),
    URL = require("url");

var request = require('request');
var _ = require('lodash');

var Entities = require('html-entities').XmlEntities;
 
var html_entities = new Entities();

var PornHub = module.exports;
var videoEmbedUrl = _.template("http://www.pornhub.com/webmasters/video_embed_code?id=<%= video_id  %>");
var videoInfoUrl = _.template("http://www.pornhub.com/webmasters/video_by_id?id=<%= video_id  %>&thumbsize=<%= size  %>");

PornHub.resolveId = function resolveId(id, cb) {
  if (typeof id !== "number") {
    return cb(Error("wrong type for id; expected number but got " + typeof id));
  }

  var get_info_url = videoInfoUrl({video_id: id, size: 'large_hd'});

  var req = http.get(get_info_url, function(res) {
    if (res.statusCode === 404) {
      return cb(Error("video not found"));
    }

    if (res.statusCode !== 301) {
      return cb(Error("incorrect status code; expected 301 but got " + res.statusCode));
    }

    return cb(null, res.headers.location);
  });

  req.once("error", cb);
};

PornHub.details = function details(url, cb) {
  var url_parsed = URL.parse(url, true);
  var video_id = url_parsed.query.viewkey;
  var url_opts = {
    video_id: video_id,
    size: 'large_hd'
  };

  var video_embed = videoInfoUrl(url_opts);
  
  var get_info_url = videoInfoUrl(url_opts);
  var get_embed_url = videoEmbedUrl(url_opts);

  request.get(get_info_url, function(err, res, body) {
      if (err) {
        return cb(err);
      }
      // console.log('res', res);
      if (res.statusCode === 404) {
        return cb(Error("pornhub: video not found"));
      }

      if (res.statusCode !== 200) {
        return cb(Error("pornhub: incorrect status code; expected 200 but got " + res.statusCode));
      }

      var data = body;
      if (typeof body === 'string') {
        data = JSON.parse(body);
      }

      // console.log('body', body);

      var title;
      title = data.video.title;

      var duration;
      duration = data.video.duration;

      var tags = [];
      for (var i = data.video.tags.length - 1; i >= 0; i--) {
        var tag = data.video.tags[i];
        tags.push(tag.tag_name);
      };

      var thumb;
      thumb = data.video.default_thumb;

      request.get(get_embed_url, function(err, res, body) {
        var data = body;
        if (typeof body === 'string') {
          data = JSON.parse(body);
        }

        // console.log("data", JSON.stringify(data, null, 2));

        var html;
        if (data.embed && data.embed.code) {
          html = html_entities.decode(data.embed.code);
        }
        return cb(null, {title: title, duration: duration, tags: tags, thumb: thumb, html: html}); 
      });

  });
};

PornHub.constructSearchUrl = function constructSearchUrl(parameters) {
  return "http://www.pornhub.com/webmasters/search?" + qs.stringify(parameters);
};

PornHub.search = function search(parameters, cb) {
  var req = http.get(this.constructSearchUrl(parameters), function(res) {
    var body = Buffer(0);

    res.on("readable", function() {
      var chunk;
      while (chunk = res.read()) {
        body = Buffer.concat([body, chunk]);
      }
    });

    if (res.statusCode !== 200) {
      return cb(Error("incorrect status code; expected 200 but got " + res.statusCode));
    }

    res.on("end", function() {
      body = body.toString("utf8");

      var $ = cheerio.load(body);

      var videos = $(".thumbBlock > .thumbInside").map(function(i, e) {
        var find;

        if ($(e).find("script").length) {
          find = cheerio.load($(e).find("script").text().replace(/^thumbcastDisplayRandomThumb\('(.+?)'\);$/, "$1"));
        } else {
          find = $(e).find.bind($(e));
        }

        return {
          url: URL.resolve("http://www.PornHub.com/", find("div.thumb > a").attr("href").replace("/THUMBNUM/", "/")),
          title: find("p > a").text(),
          duration: $(e).find("span.duration").text().replace(/[\(\)]/g, "").trim(),
        };
      });

      var total = parseInt($("h3.blackTitle").text().replace(/[\r\n]/g, " ").replace(/^.*- (\d+) results.*$/, "$1"), 10);

      return cb(null, {total: total, videos: videos});
    });
  });

  req.once("error", cb);
};


xvideos.js
var cheerio = require("cheerio"),
    http = require("http"),
    qs = require("querystring"),
    url = require("url");

var XVideos = module.exports;

XVideos.resolveId = function resolveId(id, cb) {
  if (typeof id !== "number") {
    return cb(Error("wrong type for id; expected number but got " + typeof id));
  }

  var req = http.get("http://www.xvideos.com/video" + id, function(res) {
    if (res.statusCode === 404) {
      return cb(Error("video not found"));
    }

    if (res.statusCode !== 301) {
      return cb(Error("incorrect status code; expected 301 but got " + res.statusCode));
    }

    return cb(null, res.headers.location);
  });

  req.once("error", cb);
};

XVideos.details = function details(url, cb) {
  var req = http.get(url, function(res) {
    var body = Buffer(0);

    res.on("readable", function() {
      var chunk;
      while (chunk = res.read()) {
        body = Buffer.concat([body, chunk]);
      }
    });

    if (res.statusCode === 404) {
      return cb(Error("video not found"));
    }

    if (res.statusCode !== 200) {
      return cb(Error("incorrect status code; expected 200 but got " + res.statusCode));
    }

    res.on("end", function() {
      body = body.toString("utf8");

      var $ = cheerio.load(body);

      var title = $("#main > h2").text().replace(/-[^-]+$/, "").trim();

      var tags = $("#video-tags > li > a").map(function(i, e) {
        return $(e).text().trim();
      }).filter(function(e) {
        return e !== "tags";
      });

      var duration = $(".duration").text().trim().replace(/^-/, "").trim();

      var flv;
      if (matches = body.match(/flv_url=(http%3A%2F%2F.+?)&amp;/)) {
        flv = unescape(matches[1]);
      }
      if (!flv) {
        return cb(Error("couldn't find flv"));
      }

      var thumb;
      if (matches = body.match(/url_bigthumb=(http:\/\/.+?)&amp;/)) {
        thumb = unescape(matches[1]);
      }

      var html = $("input#mediaEmbedCodeInput").attr('value');

      return cb(null, {title: title, duration: duration, tags: tags, flv: flv, thumb: thumb, html: html});
    });
  });

  req.once("error", cb);
};

XVideos.constructSearchUrl = function constructSearchUrl(parameters) {
  return "http://www.xvideos.com/?" + qs.stringify(parameters);
};

XVideos.search = function search(parameters, cb) {
  var req = http.get(this.constructSearchUrl(parameters), function(res) {
    var body = Buffer(0);

    res.on("readable", function() {
      var chunk;
      while (chunk = res.read()) {
        body = Buffer.concat([body, chunk]);
      }
    });

    if (res.statusCode !== 200) {
      return cb(Error("incorrect status code; expected 200 but got " + res.statusCode));
    }

    res.on("end", function() {
      body = body.toString("utf8");

      var $ = cheerio.load(body);

      var videos = $(".thumbBlock > .thumbInside").map(function(i, e) {
        var find;

        if ($(e).find("script").length) {
          find = cheerio.load($(e).find("script").text().replace(/^thumbcastDisplayRandomThumb\('(.+?)'\);$/, "$1"));
        } else {
          find = $(e).find.bind($(e));
        }

        return {
          url: url.resolve("http://www.xvideos.com/", find("div.thumb > a").attr("href").replace("/THUMBNUM/", "/")),
          title: find("p > a").text(),
          duration: $(e).find("span.duration").text().replace(/[\(\)]/g, "").trim(),
        };
      });

      var total = parseInt($("h3.blackTitle").text().replace(/[\r\n]/g, " ").replace(/^.*- (\d+) results.*$/, "$1"), 10);

      return cb(null, {total: total, videos: videos});
    });
  });

  req.once("error", cb);
};
